presto-hive/src/test/java/com/facebook/presto/hive/TestEventListenerWithExchangeMaterialization.java

```diff
@@ -0,0 +1,264 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.execution.StageId;
+import com.facebook.presto.spi.Plugin;
+import com.facebook.presto.spi.QueryId;
+import com.facebook.presto.spi.eventlistener.EventListener;
+import com.facebook.presto.spi.eventlistener.EventListenerFactory;
+import com.facebook.presto.spi.eventlistener.QueryCompletedEvent;
+import com.facebook.presto.spi.eventlistener.QueryCreatedEvent;
+import com.facebook.presto.spi.eventlistener.QueryMetadata;
+import com.facebook.presto.spi.eventlistener.SplitCompletedEvent;
+import com.facebook.presto.testing.MaterializedResult;
+import com.facebook.presto.tests.DistributedQueryRunner;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import org.intellij.lang.annotations.Language;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import static com.facebook.presto.hive.HiveQueryRunner.createQueryRunner;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.airlift.tpch.TpchTable.getTables;
+import static java.util.Objects.requireNonNull;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertFalse;
+import static org.testng.Assert.assertTrue;
+
+public class TestEventListenerWithExchangeMaterialization
+{
+    private EventsBuilder generatedEvents = new EventsBuilder();
+    private DistributedQueryRunner queryRunner;
+    private Session session;
+
+    @BeforeClass
+    private void setUp()
+            throws Exception
+    {
+        queryRunner = createQueryRunner(
+                getTables(),
+                new ImmutableMap.Builder<String, String>()
+                        .put("query.partitioning-provider-catalog", "hive")
+                        .put("query.exchange-materialization-strategy", "ALL")
+                        .put("experimental.runtime-optimizer-enabled", "true")
+                        .put("experimental.enable-stats-collection-for-temporary-table", "true")
+                        .put("join-distribution-type", "PARTITIONED")
```

Comment:
you should also set "optimizer.join-reordering-strategy" to "ELIMINATE_CROSS_JOINS"